packages/edit-site/src/components/block-editor/index.js

```diff
@@ -51,28 +58,252 @@ const LAYOUT = {
+				settings={ settings }
+				templateType={ templateType }
+			/>
+
+			<ReusableBlocksMenuItems />
+		</BlockEditorComponent>
+	);
+}
+
+/**
+ * Factory to isolate choosing the appropriate block editor
+ * component to handle a given template type/entity.
+ *
+ * @param {string} templateType the template (entity) type being edited
+ * @return {JSX.Element} the block editor component to use.
+ */
+function getBlockEditorComponent( templateType ) {
+	let Component = null;
+
+	switch ( templateType ) {
+		case 'wp_navigation':
+			Component = NavigationBlockEditor;
+			break;
+		default:
+			Component = DefaultBlockEditor;
+			break;
+	}
+
+	return Component;
+}
+
+/**
+ * Block editor component for editing navigation menus.
+ *
+ * Note: Navigation entities require a wrapping Navigation block to provide
+ * them with some basic layout and styling. Therefore we create a "ghost" block
+ * and provide it will a reference to the navigation entity ID being edited.
+ *
+ * In this scenario it is the **block** that handles syncing the entity content
+ * whereas for other entities this is handled by entity block editor.
+ *
+ * @param {number} navigationMenuId the navigation menu ID
+ * @return {[WPBlock[], Function, Function]} The block array and setters.
+ */
+function NavigationBlockEditor( { children, settings, templateType } ) {
+	const noop = () => {};
+
+	useNavigationFocusMode( {
+		templateType,
+	} );
+
+	const navigationMenuId = useEntityId( 'postType', 'wp_navigation' );
+
+	const blocks = useMemo( () => {
+		return [
+			createBlock( 'core/navigation', {
+				ref: navigationMenuId,
+				// As the parent editor is locked with `templateLock`, the template locking
+				// must be explicitly "unset" on the block itself to allow the user to modify
+				// the block's content.
+				templateLock: false,
+			} ),
+		];
+	}, [ navigationMenuId ] );
+
+	return (
+		<ExperimentalBlockEditorProvider
+			settings={ settings }
+			value={ blocks }
+			onInput={ noop }
+			onChange={ noop }
+			useSubRegistry={ false }
+		>
+			{ children }
+		</ExperimentalBlockEditorProvider>
+	);
+}
+
+function DefaultBlockEditor( { children, templateType, settings } ) {
+	const [ blocks, onInput, onChange ] = useEntityBlockEditor(
+		'postType',
+		templateType
+	);
+
+	return (
+		<ExperimentalBlockEditorProvider
+			settings={ settings }
+			value={ blocks }
+			onInput={ onInput }
+			onChange={ onChange }
+			useSubRegistry={ false }
+		>
+			{ children }
+		</ExperimentalBlockEditorProvider>
+	);
+}
+
+function SiteEditorCanvas( { templateType, settings } ) {
+	const { clearSelectedBlock } = useDispatch( blockEditorStore );
+	const { isViewMode, isFocusMode } = useSiteEditorMode();
+	const [ resizeObserver, sizes ] = useResizeObserver();
+
+	// Get the blocks directly from the block editor store.
+	const { blocks } = useSelect( ( select ) => {
+		const { getBlocks } = select( blockEditorStore );
+		return {
+			blocks: getBlocks(),
+		};
+	}, [] );
```

Comment:
It seems we only need the `hasBlocks` so maybe it's simpler to just return a boolean from this useSelect to avoid re-rendering the component on each block change. (Could also use getBlockCount or something)

Reply:
I converted this to the `Boolean` return form.

Reply:
Oh I just realised you said try `getBlockCount` as well... 🤔